### PR TITLE
Add ActivationPolicy= [Link] parameter

### PR DESCRIPTION
**Pull request by ddstreet**
This parameter controls when networkd will adjust the interface IFF_UP flag.  It allows always keeping the interface either up or down, or just initially taking the interface up or down at networkd startup, or not modifying IFF_UP at all, to allow fully manual control.

This is related to #15060

### CODE REVIEW

File: man/systemd.network.xml
Comment (poettering) [inline man/systemd.network.xml]: `networkd` → `<command>systemd-networkd</command>`

File: man/systemd.network.xml
Comment (poettering) [inline man/systemd.network.xml]: ditto (and everywhere else too)

(Yeah, I know we are not very good at this right now, but let's try to get this right at least for new documentation)

File: man/systemd.network.xml
Comment (poettering) [inline man/systemd.network.xml]: previous state → state configured with this option

Comment (poettering) [reply man/systemd.network.xml]: "state bounce"? Is that established terminology? I at least (as a non-native English speaker) don#t know the term... Maybe call this "brief state change"

File: man/systemd.network.xml
Comment (poettering) [inline man/systemd.network.xml]: maybe introduce an explicit state for this called "bound" or so? And BindCarrier= would imply that the state is set to "bound"?

This would be useful when querying the state, as then BindCarrier= would be easily discoverable just by looking at the state

**Comment by ddstreet** [comment]
> I am a bit concerned about the naming conflict with the D-Bus property "AdministrativeState".
> 
> I think this new added feature probably should be called "State" anyway, but "Policy" or so. Aftr all state is what is influenced by this setting, it declares when to change state, but the state itself is independent.
> 
> Maybe call this ActivationPolicy=?

policy would make more sense for the naming, though 'activation' doesn't really feel right, though 'AdministrativePolicy' doesn't really sound much better either.

I'll address your other comments and update the name to ActivationPolicy, unless I can think of anything else to suggest :)

File: man/systemd.network.xml
Comment (ddstreet) [reply man/systemd.network.xml]: ack, changed

File: man/systemd.network.xml
Comment (ddstreet) [reply man/systemd.network.xml]: i reworded this to try to make it clearer, and dropped the 'bounce', let me know if it's better

File: man/systemd.network.xml
Comment (ddstreet) [reply man/systemd.network.xml]: great idea! i updated it to include a 'bound' state that is required/forced with BindCarrier=

**Comment by ddstreet** [comment]
> > Maybe call this ActivationPolicy=?

yep, i changed it to ActivationPolicy=, it sounds ok to me now :)

the possible values are start-up, always-up, manual, always-down, start-down, and bound; however i wonder if the 'start-up' and 'start-down' should just be 'up' and 'down'? the 'start-' seems redundant, maybe. i can change it and re-push if you think 'up' and 'down' sound better than 'start-up' and 'start-down'.

**Comment by TriMoon** [comment]
@ddstreet 
> the possible values are start-up, always-up, manual, always-down, start-down, and bound; however i wonder if the 'start-up' and 'start-down' should just be 'up' and 'down'? the 'start-' seems redundant, maybe. i can change it and re-push if you think 'up' and 'down' sound better than 'start-up' and 'start-down'.

IMHO, a better and more to-the-point naming would be "**initial-up**" and "**initial-down**" as that is what they are supposed to do right?
Those names also make more sense with the "Policy" rename....

**Comment by ddstreet** [comment]
> @ddstreet
> 
> > the possible values are start-up, always-up, manual, always-down, start-down, and bound; however i wonder if the 'start-up' and 'start-down' should just be 'up' and 'down'? the 'start-' seems redundant, maybe. i can change it and re-push if you think 'up' and 'down' sound better than 'start-up' and 'start-down'.
> 
> IMHO, a better and more to-the-point naming would be "**initial-up**" and "**initial-down**" as that is what they are supposed to do right?
> Those names also make more sense with the "Policy" rename....

initial-up/initial-down do sound better and more accurate than start-up/start-down, but just plain 'up' and 'down' still seem better to me (and the remaining possible values would still be 'always-up', 'always-down', 'manual' and 'bound'), though i have no strong feelings; i'll wait for @poettering opinion before changing the PR to any new names.

**Comment by poettering** [comment]
BTW, You can add "Fixes: #3031" to your commits

File: src/network/networkd-link.c
Comment (keszybz) [inline src/network/networkd-link.c]: Maybe `\"always-off\"` here.

File: src/network/networkd-link.c
Comment (keszybz) [inline src/network/networkd-link.c]: `\"always-on\"`

File: man/systemd.network.xml
Comment (keszybz) [inline man/systemd.network.xml]: Hmm, networkd now supports configuration reloads, so it may bring interfaces up not only at start. So instead of referring to networkd startup, say "... will set the administrative state of the link to up or down after configuring the interface, but will not automatically change the state afterwards".

(Edit: I added "automatically" because of c30ffcee9b4960987bc8cf201138aaf0d7b26b14.)

Also, with "start-up", if the interface is manually set to down, will `networkctl reconfigure` bring it back up? I think it should.

File: man/systemd.network.xml
Comment (keszybz) [inline man/systemd.network.xml]: I think this para should be moved before the previous one.

**Review by keszybz — COMMENTED** [review]
IIUC, this changes the behaviour slightly: the new default is "start-up", and if I'm reading the diff correctly, the code path that was taken previously corresponds to the "always-up" option. Can you please clarify (in the commit message) if there's any behaviour change.

**Comment by ddstreet** [comment]
Sorry for the delay in responding. tl;dr: i'll update the commit msg to clarify if this changes previous/default behavior.

> IIUC, this changes the behaviour slightly: the new default is "start-up", and if I'm reading the diff correctly, the code path that was taken previously corresponds to the "always-up" option. Can you please clarify (in the commit message) if there's any behaviour change.

So, this is one reason I originally felt that 'ActivationPolicy' was not as clear as 'AdministrativeState'.  It's true that the previous policy had systemd-networkd setting the IFF_UP flag bit to 1 (if not already 1) every time link_joined() is called, however, once IFF_UP is set to 1, it should *never* change back to 0, unless some administrator explicitly sets it to 0. Losing carrier won't set IFF_UP to 0, nor will any address or routing changes. The IFF_UP flag is essentially like the physical key to a car; the car doesn't necessarily start and/or run once you turn the key 'on', but it *definitely* will not start/run without the key 'on'.

As the previous behavior was to set IFF_UP to 1 anytime link_joined() is called, it doesn't exactly match the 'always-up' setting; previously, if an admin set IFF_UP to 0, systemd-networkd would notice the flag change, but do nothing; then since losing IFF_UP causes carrier to be lost, systemd-networkd would also notice carrier loss and (unless IgnoreCarrierLoss=true) would unconfigure the interface. However, since carrier would *never* be regained until some adimn manually set IFF_UP back to 1, systemd-networkd would never do anything else for the interface, and once the admin sets IFF_UP back to 1, systemd-networkd doesn't need to set IFF_UP to 1 itself (since the admin just did it manually).

However; with the new addition of 'networkctl up/down', i should take a look at this again.

File: src/network/networkd-link.c
Comment (ddstreet) [reply src/network/networkd-link.c]: ack, will change this, thanks.

**Comment by yuwata** [comment]
Sorry, I made a large change on networkd. Could you rebase this?

**Comment by ddstreet** [comment]
I split the first patch from this out, as it is fairly simple and can be considered separately, in #17392 (since it's still needed in this PR, it is still included here too)

I also dropped the patch that just added a comment about the 'AdministrativeState' dbus event, as that's not really needed/important now that this is named 'ActivationPolicy' (though I do still think the dbus event is named confusingly, at least to people familiar with the older networking 'administrative state' terminology)

I rebased and pushed just so it will be up to date, but I want to review the changes again, and I might need to adjust and push again

File: src/network/networkd-link.h
Comment (poettering) [inline src/network/networkd-link.h]: i'd appreciate a quick comment added here that explains when this is set precisely

**Review by poettering — CHANGES_REQUESTED** [review]
I think @keszybz' comments werent' addressed yet.

File: src/network/networkd-link.h
Comment (ddstreet) [reply src/network/networkd-link.h]: yep sorry, will update again today to address this and the other previous comments

File: man/systemd.network.xml
Comment (ddstreet) [reply man/systemd.network.xml]: I updated the doc, to include clarifying it 'up' and 'down' will apply only during (re)configuration; let me know if you think the new wording should be changed.

> Also, with "start-up", if the interface is manually set to down, will networkctl reconfigure bring it back up? I think it should.

I agree; I moved the flag out of the link and into the link->network, so up/down policy should be applied each reconfiguration.

File: man/systemd.network.xml
Comment (ddstreet) [reply man/systemd.network.xml]: ack, i moved it to "When <literal>up</literal> (the default)"

**Comment by ddstreet** [comment]
Ok I think I addressed previous comments, and as previously mentioned I did rename 'start-up/down' to just 'up/down', which I think makes more sense. I also updated the commit message to clarify that the default behavior in handling the IFF_UP flag is changed, to only set it once for each (re)configuration, instead of every time link_joined() was called.

**Comment by poettering** [comment]
lgtm now. I guess @yuwata might want to be the one doing a final review/merge though.

**Comment by yuwata** [comment]
Hmm, all CentOS CIs are unhappy. Also, could you rebase this, as I merged #17392?

**Comment by ddstreet** [comment]
hmm, well systemd-networkd-tests.py is failing a few tests now, with systemd-networkd-wait-online timing out. I'll run the tests locally to see what i added that's causing the failures.

**Comment by ddstreet** [comment]
last push was just a rebase, i'm still looking at what's causing the failures

File: man/systemd.network.xml
Comment (ddstreet) [reply man/systemd.network.xml]: moving the ->activated flag into link->network was a mistake on my part (since ->network can be shared between links and shouldn't hold link state, i'm pretty sure this was causing the test failures), but i just moved it back to link->activated and it should still be reset on networkctl reload, since that will call link_reconfigure() with force=True.

**Comment by ddstreet** [comment]
ok i think it was failing because i had the per-link activation flag in the link->network, which it should not have been, i moved it back to link->activated so it should correctly be a per-link flag now. Also 'networkctl reconfigure' should still work to re-apply the activation policy.

**Comment by sil2100** [comment]
Hello! I wanted to poke about the status of this PR. CI wise it seems to look good - there is one bionic-i386 CI failure which seems to be a timeout issue, possibly 'working' after a re-run. Are there any concerns regarding the design or implementation? Is there anything that can be done to get it merged? Thank you!

**Comment by yuwata** [comment]
@sil2100 v247-rc1 was released, and I guess we can release -rc2 near future. So, I'd like to make final review after v247-final released.

File: src/libsystemd/sd-network/sd-network.c
Comment (poettering) [inline src/libsystemd/sd-network/sd-network.c]: hmm, if the field is missing I'd synthesize "UP" here, i.e. the only supported policy in the status quo ante.

i mean, this doesn't matter, as we don't really support mixing client and server, but it might still be nice to add, because it's easy

**Comment by poettering** [comment]
Still looks good to me, btw. Still leaving the final merge to @yuwata though

File: src/network/networkd-network.c
Comment (yuwata) [inline src/network/networkd-network.c:340]: I'd like to set `ACTIVATION_POLICY_UP` here. If we do not need to detect a setting is actually specified or not, then we should set the default value here.

File: src/network/networkd-network.c
Comment (yuwata) [inline src/network/networkd-network.c]: And then set `ACTIVATION_POLICY_UP`.

File: src/network/networkd-network.c
Comment (yuwata) [inline src/network/networkd-network.c:251]: So, we can drop this.

File: src/network/networkd-network.c
Comment (yuwata) [inline src/network/networkd-network.c]: Not a matter, but I'd like to align these entries.

File: test/test-network/systemd-networkd-tests.py
Comment (yuwata) [inline test/test-network/systemd-networkd-tests.py]: How about to use `self.subTest()`?

File: src/network/networkctl.c
Comment (yuwata) [inline src/network/networkctl.c]: Doubled white space after comma. Is it intended to 'align' the field with the above?

**Review by yuwata — COMMENTED** [review]
LGTM. Some minor comments. Not necessary to be addressed if you do not agree with.

File: src/libsystemd/sd-network/sd-network.c
Comment (ddstreet) [reply src/libsystemd/sd-network/sd-network.c]: ack, i'll update in the next push, thanks

File: src/network/networkd-network.c
Comment (ddstreet) [reply src/network/networkd-network.c:340]: the only issue with setting it directly is in the check for BindCarrier:

```
        if (!strv_isempty(network->bind_carrier)) {                                                                                                                                                                                                                           
                if (!IN_SET(network->activation_policy, _ACTIVATION_POLICY_INVALID, ACTIVATION_POLICY_BOUND))                                                                                                                                                                 
                        log_warning("%s: ActivationPolicy=bound is required with BindCarrier=. "                                                                                                                                                                              
                                    "Setting ActivationPolicy=bound.", network->filename);                                                                                                                                                                                    
                network->activation_policy = ACTIVATION_POLICY_BOUND;                                                                                                                                                                                                         

```

that currently avoids issuing the conflict warning if ActivationPolicy was unset; but if we set it to its default, we can no longer tell if it was unset or if it was set to 'up', so the warning would either need to be suppressed when someone configures both BindCarrier= and ActivationPolicy=up, or the warning would need to be logged (incorrectly) when someone configures only BindCarrier=.

Do you think we should just drop the warning completely?

File: src/network/networkd-network.c
Comment (ddstreet) [reply src/network/networkd-network.c]: ack

File: src/network/networkd-network.c
Comment (ddstreet) [reply src/network/networkd-network.c]: ack

File: test/test-network/systemd-networkd-tests.py
Comment (ddstreet) [reply test/test-network/systemd-networkd-tests.py]: it does use subTest() in the up/down/up/down iterations further down in the test, but let me check if it's possible to nest subTests

File: src/network/networkctl.c
Comment (ddstreet) [reply src/network/networkctl.c]: wow, good eye! I don't think I intended it for alignment, I'm pretty sure it was just a typo :) will fix in the next push

File: src/network/networkd-network.c
Comment (yuwata) [reply src/network/networkd-network.c:340]: Ah, indeed. Please ignore my comment here.

**Comment by ddstreet** [comment]
pushed with rebase and some updates as requested by feedback; I think only remaining feedback I'm working on is:
-whether to set 'up' directly instead of using INVALID value, and what to do about bindcarrer conflict warning
-using (nested) subtest in testcase

File: test/test-network/systemd-networkd-tests.py
Comment (ddstreet) [reply test/test-network/systemd-networkd-tests.py]: the nested subTest() calls seem to work great, thanks for the suggestion!

**Comment by ddstreet** [comment]
I think this addresses all the remaining feedback, if the ci passes PTAL

**Pull request merged**
